From e15ce05250a47248190c20d927882525d3750a66 Mon Sep 17 00:00:00 2001
From: Brian Delwiche <delwiche@google.com>
Date: Mon, 29 Jun 2026 13:21:30 -0700
Subject: [PATCH 09/10] Fix out-of-bounds heap write in SnoopLogger

If specific Developer Options are enabled (such as SnoopLogger filtering active with the PBAP/MAP filter profile set to "magic"), a malicious Bluetooth controller can send a short HCI ACL packet that triggers a heap out-of-bounds write in `SnoopLogger::PayloadStrip()`. This occurs because a fixed string ('PROHIBITED') is written to the payload, and if the packet is extremely small (e.g., 8 bytes), the heap-allocated buffer does not have sufficient capacity.

A safeguard in `FilterCapturedPacket` intended to expand the buffer for small packets was logically inverted, meaning short packets were incorrectly skipped from being resized.

This commit fixes the issue by changing the guard condition from `>` to `<=`. This ensures that small packets correctly receive an `EXTRA_BUF_SIZE` padding, providing enough underlying buffer space for the "PROHIBITED" string overwrite and preventing the out-of-bounds heap write.

Test: alint against snoop_logger.cc; m bluetooth_test net_test_bluetooth bluetooth_packet_parser_test

Bug: 522372274
Flag: EXEMPT BUGFIX
This was autogenerated by MendIt (go/androidmendit)

Change-Id: I3fae8e1b8af9f05970de542317bcdcc3d020bf9e
(cherry picked from commit 0083751f4ce0a648674ee6e31b480bafa2e0e97e)
---
 system/gd/hal/snoop_logger.cc      |  4 +++-
 system/gd/hal/snoop_logger_test.cc | 12 ++++++++----
 2 files changed, 11 insertions(+), 5 deletions(-)

diff --git a/system/gd/hal/snoop_logger.cc b/system/gd/hal/snoop_logger.cc
index 42203697572..25f52f8274c 100644
--- a/system/gd/hal/snoop_logger.cc
+++ b/system/gd/hal/snoop_logger.cc
@@ -1132,7 +1132,7 @@ void SnoopLogger::FilterCapturedPacket(HciPacket& packet, Direction direction, P
       IsFilterEnabled(kBtSnoopLogFilterProfileMapModeProperty)) {
     // If HeadersFiltered applied, do not use ProfilesFiltered
     if (length == ntohl(header.length_original)) {
-      if (packet.size() + EXTRA_BUF_SIZE > DEFAULT_PACKET_SIZE) {
+      if (packet.size() + EXTRA_BUF_SIZE <= DEFAULT_PACKET_SIZE) {
         // Add additional bytes for magic string in case
         // payload length is less than the length of magic string.
         packet.resize((size_t)(packet.size() + EXTRA_BUF_SIZE));
@@ -1152,6 +1152,8 @@ void SnoopLogger::FilterCapturedPacket(HciPacket& packet, Direction direction, P
       length = L2CAP_HEADER_SIZE + PACKET_TYPE_LENGTH;
     }
   }
+
+  length = std::min(length, static_cast<uint32_t>(packet.size() + PACKET_TYPE_LENGTH));
 }
 
 void SnoopLogger::Capture(const HciPacket& immutable_packet, Direction direction, PacketType type) {
diff --git a/system/gd/hal/snoop_logger_test.cc b/system/gd/hal/snoop_logger_test.cc
index c54601c769d..38521fe7c2a 100644
--- a/system/gd/hal/snoop_logger_test.cc
+++ b/system/gd/hal/snoop_logger_test.cc
@@ -111,7 +111,11 @@ static const std::vector<bluetooth::hal::HciPacket> kTestData = {
         {0x02, 0x20, 0x11, 0x00, 0x0d, 0x00, 0x41, 0x00, 0x9d, 0xef, 0x43,
          0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00},
         {0x02, 0x00, 0x12, 0x00, 0x0e, 0x00, 0x40, 0x00, 0x9f, 0xff, 0x11,
-         0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00}};
+         0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00},
+        // Short packet (incoming) - L2CAP CID: 0x0041 (local_cid)
+        {0x02, 0x20, 0x04, 0x00, 0x00, 0x00, 0x41, 0x00},
+        // Short packet (outgoing) - L2CAP CID: 0x0040 (remote_cid)
+        {0x02, 0x00, 0x04, 0x00, 0x00, 0x00, 0x40, 0x00}};
 
 }  // namespace
 
@@ -960,7 +964,7 @@ TEST_F(SnoopLoggerTest, profiles_filtered_pbap_magic_test) {
               (filterMapModeProperty->find(SnoopLogger::kBtSnoopLogFilterProfileModeMagic) !=
                std::string::npos));
 
-  auto snoop_logger = NewSnoopLogger(15, SnoopLogger::kBtSnoopLogModeFiltered, false, false);
+  auto snoop_logger = NewSnoopLogger(20, SnoopLogger::kBtSnoopLogModeFiltered, false, false);
 
   snoop_logger->SetL2capChannelOpen(conn_handle, local_cid, remote_cid, psm, false);
   snoop_logger->SetRfcommPortOpen(conn_handle, local_cid, dlci, profile_uuid_pbap, flow);
@@ -1025,7 +1029,7 @@ TEST_F(SnoopLoggerTest, profiles_filtered_pbap_header_test) {
               (filterMapModeProperty->find(SnoopLogger::kBtSnoopLogFilterProfileModeHeader) !=
                std::string::npos));
 
-  auto snoop_logger = NewSnoopLogger(15, SnoopLogger::kBtSnoopLogModeFiltered, false, false);
+  auto snoop_logger = NewSnoopLogger(20, SnoopLogger::kBtSnoopLogModeFiltered, false, false);
 
   snoop_logger->SetL2capChannelOpen(conn_handle, local_cid, remote_cid, psm, false);
   snoop_logger->SetRfcommPortOpen(conn_handle, local_cid, dlci, profile_uuid_pbap, flow);
@@ -1090,7 +1094,7 @@ TEST_F(SnoopLoggerTest, profiles_filtered_pbap_fullfilter_test) {
               (filterMapModeProperty->find(SnoopLogger::kBtSnoopLogFilterProfileModeFullfillter) !=
                std::string::npos));
 
-  auto snoop_logger = NewSnoopLogger(15, SnoopLogger::kBtSnoopLogModeFiltered, false, false);
+  auto snoop_logger = NewSnoopLogger(20, SnoopLogger::kBtSnoopLogModeFiltered, false, false);
 
   snoop_logger->SetL2capChannelOpen(conn_handle, local_cid, remote_cid, psm, false);
   snoop_logger->SetRfcommPortOpen(conn_handle, local_cid, dlci, profile_uuid_pbap, flow);
-- 
2.55.0

